feat: [BE] 백엔드 개발 환경 설정

youngxpepp opened the pull request:
## 📕 제목

백엔드 개발 환경 설정

## 📗 작업 내용

> 구현 내용 및 작업 했던 내역

- [x] ESLint, Prettier 설정
- [x] ES 최신 문법 빌드를 위한 babel 설정
- [x] Application과 ApplicationFactory를 통한 서버 어플리케이션 초기화 작업
- [x] TypeORM을 통해 데이터베이스 연결 설정
- [x] 예제 엔티티 User 생성
- [x] 공통된 에러 처리를 위한 미들웨어 개발
- [x] 로컬에서 실행하기 위한 MySQL 도커 컴포즈 생성
- [x] 서버 실행을 위한 다양한 scripts 명령어 (package.json 참조)
- [x] 개발 환경에 따른 환경변수 세분화

## 📘 PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점

- 관련 이슈 #4 
- Express에 관련된 설정은 backend/Application.js의 생성자 부분을 확인해주세요.
- TypeORM 예제 엔티티를 선언했습니다. 참고해주세요. backend/src/model/user.js
- backend/src/common/config/database/connection-option-generator.js의 14번 라인과 같이 엔티티를 등록해주어야 합니다.
- TypeORM에 관련된 문서는 https://typeorm.io/#/ 공식 문서를 확인해주세요.
- 환경변수 설정은 위키에 올리겠습니다✌️ (https://github.com/boostcamp-2020/IssueTracker-40/wiki/%EB%B0%B1%EC%97%94%EB%93%9C-%ED%99%98%EA%B2%BD-%EB%B3%80%EC%88%98-%EA%B0%80%EC%9D%B4%EB%93%9C))
- npm run start, npm run start:dev는 절대 프로덕션용이 아닙니다. 프로덕션에서 사용하기 위해선 npm run build -> npm run start:prod를 순차적으로 실행해야 합니다. 빌드 결과물은 backend/dist 경로에 생성됩니다.
- docker-compose가 설치되어 있다면, `docker-compose up -d` 라는 터미널 명령어로 MySQL 5.7을 컨테이너로 실행할  수 있습니다. 해당 컨테이너는 로컬 환경에서의 테스트 용이성을 위해 제작되었고, MySQL 사용이 어렵다면 SQLite3를 사용 가능합니다.

youngxpepp committed:
feat: [BE] 백엔드 개발 환경 설정

- ESLint, Prettier 설정
- ES 최신 문법 빌드를 위한 babel 설정
- Application과 ApplicationFactory를 통한 서버 어플리케이션 초기화 작업
- TypeORM을 통해 데이터베이스 연결 설정
- 예제 엔티티 User 생성
- 공통된 에러 처리를 위한 미들웨어 개발
- 로컬에서 실행하기 위한 MySQL 도커 컴포즈 생성
- 서버 실행을 위한 다양한 scripts 명령어 (package.json 참조)
- 개발 환경에 따른 환경변수 세분화 (상세 환경변수는 위키 참조)

Do-ho commented on backend/docker-compose.yml:
```diff
@@ -0,0 +1,23 @@
+    ports:
+      - 3307:3306
+    environment:
+      MYSQL_DATABASE: "issuetrackerdb"
+      MYSQL_USER: "team40"
+      MYSQL_PASSWORD: "123123"
+      MYSQL_ROOT_PASSWORD: "root"
```
이 부분이 하드 코딩으로 들어간 것 같습니다. 무조건적으로 하드코딩으로 들어가야할까요?

youngxpepp replied:
환경변수 설정해보겠습니다!
https://docs.docker.com/compose/environment-variables/

Do-ho commented on backend/src/common/error/error-code.js:
```diff
@@ -0,0 +1,7 @@
+const ErrorCode = {
+    ENTITY_NOT_FOUND: { code: 1000, httpStatusCode: 404, message: "Entity is not found" },
+    ENTITY_ALREADY_EXIST: { code: 1001, httpStatusCode: 400, message: "Entity already exists" },
+    AUTHENTICATION_FAILED: { code: 2001, httpStatusCode: 400, message: "Authentication is failed" }
```
AUTHENTICATION_FAILED httpStatusCode가 401이 되어야할 것 같습니다!
```js
AUTHENTICATION_FAILED: { code: 2001, httpStatusCode: 401, message: "Authentication is failed" }
```

youngxpepp replied:
넵 조언 감사합니다

Do-ho commented on backend/src/main.js:
```diff
@@ -0,0 +1,8 @@
+import { ApplicationFactory } from "./application-factory";
+
+async function main() {
+    const app = await ApplicationFactory.create();
```
설명 들었었지만 정말 좋은 방식인 것 같아요 😄

Do-ho commented on backend/src/model/user.js:
```diff
@@ -0,0 +1,21 @@
+import { Column, CreateDateColumn, Entity, PrimaryGeneratedColumn, UpdateDateColumn } from "typeorm";
+
+@Entity({ name: "user" })
+class User {
+    @PrimaryGeneratedColumn("increment", { type: "int" })
+    id;
+
+    @Column({ name: "email", type: "varchar" })
+    email;
+
+    @Column({ name: "name", type: "varchar" })
+    name;
+
+    @CreateDateColumn({ name: "created_at", type: "datetime" })
+    createdAt;
+
+    @UpdateDateColumn({ name: "updated_at", type: "datetime" })
+    updatedAt;
+}
+
+export { User };
```
model 너무 깔끔하게 작성되는 것 같습니다. 역시  typeORM 👍

Do-ho reviewed (commented):
건홍님! 백엔드 개발 환경 설정하느라 정말 고생하셨습니다.
몇 가지 의문점을 PR로 남겨요!

wooojini commented on backend/src/common/env/database-env.js:
```diff
@@ -0,0 +1,65 @@
+    @IsNotEmpty()
+    @IsString()
+    databaseLogging;
+
```
class-validator를 사용하니 클래스를 작성할 때 견고하게 작성되는 것 같아 좋습니다 👍

wooojini commented on backend/src/common/error/error-code.js:
```diff
@@ -0,0 +1,7 @@
+    ENTITY_NOT_FOUND: { code: 1000, httpStatusCode: 404, message: "Entity is not found" },
+    ENTITY_ALREADY_EXIST: { code: 1001, httpStatusCode: 400, message: "Entity already exists" },
+    AUTHENTICATION_FAILED: { code: 2001, httpStatusCode: 400, message: "Authentication is failed" }
+};
```
고생하셨습니다!! 하나의 BusinessError에서 에러코드에 따라 선택을 하면 에러를 호출하는 쪽에서 에러코드에 대한 객체도 가져와서 사용을 하게 될 것 같습니다. 에러코드에 해당하는 부분을 별도의 에러 클래스로 정의하는 방식은 어떨까요?? 여러분들의 의견이 궁급합니다!!!

이런식으로 바로 에러만 호출할 수 있으면 어떨까요?
```
  new EntityNotFoundError()
```

Do-ho replied:
우진님 말이 정말 좋은 것 같습니다. 저도 this.상태로 들어가면 좋겠네요.

youngxpepp replied:
넵 BusinessError를 상속해서 만들 수 있도록 바꿔볼게요!

wooojini reviewed (commented):
건홍님이 계서서 저희 프로젝트의 벡엔드가 단단해지는 것을 느겼습니다!

jinhyukoo commented on backend/src/model/user.js:
```diff
@@ -0,0 +1,21 @@
+    updatedAt;
+}
+
+export { User };
```
아마 나중에 처리하겠지만 email이나 name에 Allow Not Null 같은 설정도 추가되어야 할 것 같다는 생각이 드네요! model 정의는 정말 직관적이네요~ 너무 좋습니다~~~

youngxpepp replied:
넵 필요한 경우 class-validator의 여러 유효성 검사를 추가해주면 될 것 같습니다!

jinhyukoo reviewed (approved):
건홍님 덕에 견고한 백엔드 환경 설정이 완료되었네요! 정말 고생 많으셨습니다!!! 😊

Do-ho reviewed (approved):
Approve 하겠습니다 :)

youngxpepp committed:
feat: [BE] docker-compose 환경변수 분리

- docker-compose.yml에 있는 설정을 외부에서 주입 가능하게끔 분리함

youngxpepp committed:
feat: 에러를 발생시키는 로직 변경

- BusinessError에 직접 ErrorCode를 넣게 되면 사용이 번거로움
- BusinessError를 상속하는 에러를 생성하여 부모 생성자에 ErrorCode를 넣는 방식으로 변경

youngxpepp commented:
이곳에 도커 환경변수 설정도 추가했습니다! 도커를 실행할 땐 다음과 같은 명령어로 실행해주세요.
`docker-compose --env-file .env.docker-compose up -d`
https://github.com/boostcamp-2020/IssueTracker-40/wiki/%EB%B0%B1%EC%97%94%EB%93%9C-%ED%99%98%EA%B2%BD-%EB%B3%80%EC%88%98-%EA%B0%80%EC%9D%B4%EB%93%9C

wooojini reviewed (approved):
논의한데로 깔끔하게 바뀌었네요!! 고생하셨습니다!!
